src/optimizer/rule/IndexScanRule.cpp

```diff
@@ -0,0 +1,542 @@
+           ? createIQCWithLogicAnd(iqctx, items, qctx, groupExpr)
+           : createIQCWithLogicOR(iqctx, items, qctx, groupExpr);
+}
+
+Status IndexScanRule::createIQCWithLogicAnd(IndexQueryCtx &iqctx,
+                                            const FilterItems& items,
+                                            graph::QueryContext *qctx,
+                                            const OptGroupExpr *groupExpr) const {
+    auto index = findOptimalIndex(qctx, groupExpr, items);
+    if (index == nullptr) {
+        return Status::IndexNotFound();
+    }
+
+    return appendIQCtx(index, items, iqctx);
+}
+
+Status IndexScanRule::createIQCWithLogicOR(IndexQueryCtx &iqctx,
+                                           const FilterItems& items,
+                                           graph::QueryContext *qctx,
+                                           const OptGroupExpr *groupExpr) const {
+    for (auto const& item : items.items) {
+        auto index = findOptimalIndex(qctx, groupExpr, FilterItems({item}));
+        if (index == nullptr) {
+            return Status::IndexNotFound();
+        }
+        auto ret = appendIQCtx(index, FilterItems({item}), iqctx);
+        NG_RETURN_IF_ERROR(ret);
+    }
+    return Status::OK();
+}
+
+Status IndexScanRule::appendIQCtx(const IndexItem& index,
+                                  const FilterItems& items,
+                                  IndexQueryCtx &iqctx) const {
+    auto fields = index->get_fields();
+    IndexQueryContext ctx;
+    decltype(ctx.column_hints) hints;
+    for (const auto& field : fields) {
+        bool found = false;
+        FilterItems fieldItems;
+        for (const auto& item : items.items) {
+            if (std::get<0>(item) != field.get_name()) {
+                continue;
+            }
+            fieldItems.addItem(std::get<0>(item), std::get<1>(item), std::get<2>(item));
+            found = true;
+        }
+        if (!found) break;
+        // TODO (sky) : rewrite filter expr. NE expr should be add filter expr .
+        auto it = std::find_if(fieldItems.items.begin(), fieldItems.items.end(),
+                               [](const auto &ite) {
+                                   return std::get<1>(ite) == RelationalExpression::Kind::kRelNE;
+                               });
+        if (it != fieldItems.items.end()) {
+            break;
+        }
+        auto ret = appendColHint(hints, fieldItems, field);
+        NG_RETURN_IF_ERROR(ret);
+    }
+    ctx.set_index_id(index->get_index_id());
+    // TODO (sky) : rewrite expr and set filter
+    ctx.set_column_hints(std::move(hints));
+    iqctx->emplace_back(std::move(ctx));
+    return Status::OK();
+}
+
+#define CHECK_BOUND_VALUE(v)                         \
+if (v == Value(NullType::BAD_TYPE)) {                \
+    LOG(ERROR) << "Encode value error . field : "    \
+           << col.get_name();                        \
+    return Status::Error();                          \
+}
```

Comment:
It's better to wrap this macro with `do { xxx; } while(0)`, if so we can add semicolon after `CHECK_BOUND_VALUE` and make compiler happy, like:

```
CHECK_BOULD_VALUE(value);
```

And i notice that the used `col` is not passed in by parameter and depends on the macro context, if so, when we changed col outside,  some failure will be resulted in.

Reply:
make sense, thanks.

Reply:
if (v == Value::kNullBadType)